### PR TITLE
Implemented necessary signaling for Shutdown() and ShutDownWithAbort()

### PR DESCRIPTION
**Pull request by troutkilroy**
While the .NET Apache library is compiled to use DedicatedTaskRunner, we prefer to compile from source and enable PooledTaskRunner (as set in TaskRunnerFactory.cs). However PooledTaskRunner.cs has an obvious issue with shutdown() as it will block on an infinite sleep if the task is busy (i.e., iterating == true). Not sure what was the intent of the sleep() calls, but the Java version of this code we believe is correct with event notification to control shutdown of the task. This change basically brings the C# implementation in line with that using Monitor (similar to what was done with the C# implementation of DedicatedTaskRunner).

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
@Havret if youre doing a respin, do you think its worth adding this, or it should wait till after?

**Comment by michaelandrepearce** [comment]
@troutkilroy i was about to look to merge this, but noticed there is no test case, can you add a test case for this please. 
a) to demonstrate the issue
b) prove that this fix does fix the issue
c) ensure the issue is not regressed by mistake

File: src/Threads/PooledTaskRunner.cs [hunk 1]
```diff
@@ -106,34 +106,34 @@ public void Shutdown(TimeSpan timeout)
...
-                    {
-                        Thread.Sleep(timeout);
-                    }
-
-                    if(iterating)
-                    {
-                        runningThread.Abort();
+          if(iterating)
+          {
+						Monitor.Wait(runable, timeout);
```
Comment (michaelpearce-gain) [inline src/Threads/PooledTaskRunner.cs:125]: formatting seems to a bit over the place

**Comment by troutkilroy** [comment]
Sorry for late response. Yes I created a test case for this, and it will demonstrate failure with the current code and success with fix,. Also I have a correction to the code I submitted. We can cancel this pull, and I'll submit a new one. 

One question though. For the test case I added  PooledTaskRunnerTest.cs file to tests/threads. However unlike the other taskrunner implementations the PooledTaskRunner class is not declared public, so I either make it public or I'll have to add an assembly config to the main project with an InternalsVisibleTo attribute pointing at the test assembly. Making it public would be easier. Which do you prefer?

**Comment by michaelandrepearce** [comment]
I have no issue if you make it public

**Comment by michaelpearce-gain** [comment]
@troutkilroy are you re-submitting? Or shall i close the issue out?

**Pull request closed without merging**